hashmap/src/main.rs

```diff
@@ -0,0 +1,273 @@
+use std::collections::HashMap;
+
+//mod tej_serializer;
+pub trait MeasurementConsumer {
+    type Error;
+    type Data;
+    fn start(&mut self) -> Result<(), Self::Error>;
+    fn end(&mut self) -> Result<(), Self::Error>;
+    fn timestamp(&mut self, value: String) -> Result<(), Self::Error>;
+    fn measurement(
+        &mut self,
+        group: Option<&str>,
+        name: &str,
+        value: f64,
+    ) -> Result<(), Self::Error>;
+    fn start_group(&mut self, group: &str) -> Result<(), Self::Error>;
+    fn end_group(&mut self) -> Result<(), Self::Error>;
+}
+#[derive(Debug)]
+pub struct MeasurementMap {
+    pub timestamp: String,
+    pub values: HashMap<String, Measurement>,
+}
+#[derive(Debug)]
+pub enum Measurement {
+    Single(f64),
+    Multi(HashMap<String, f64>),
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum ThinEdgeJsonMapError {
+    #[error("Duplicated measurement: {0}")]
+    DuplicatedMeasurement(String),
+
+    #[error("Duplicated measurement: {0}.{1}")]
+    DuplicatedSubMeasurement(String, String),
+}
+
+impl MeasurementMap {
+    fn new() -> Self {
+        Self {
+            timestamp: "4-20-2021".into(),
+            values: HashMap::new(),
+        }
+    }
+
+    fn serialize_tej(&self) -> Vec<u8> {
+        let mut t_serializer = ThinEdgeJsonSerializer::new();
+        t_serializer.start();
+        t_serializer.timestamp(self.timestamp.clone());
+        for (key, value) in self.values.iter() {
+            match value {
+                Measurement::Single(sv) => {
+                    t_serializer.measurement(None, key, *sv);
+                }
+                Measurement::Multi(m) => {
+                    t_serializer.start_group(key);
+                    for (key, value) in m.iter() {
+                        t_serializer.measurement(None, key, *value);
+                    }
+                    t_serializer.end_group();
+                }
+            }
+        }
+        //Prints the serialized TEJ Message
+        t_serializer.end();
+        t_serializer.tej_serailized()
+    }
+}
+
+impl MeasurementConsumer for MeasurementMap {
+    type Error = ThinEdgeJsonMapError;
+    type Data = MeasurementMap;
+
+    fn start(&mut self) -> Result<(), Self::Error> {
+        Ok(())
+    }
+
+    fn end(&mut self) -> Result<(), Self::Error> {
+        Ok(())
+    }
+
+    fn timestamp(&mut self, value: String) -> Result<(), Self::Error> {
+        self.timestamp = value;
+        Ok(())
+    }
+
+    fn measurement(
+        &mut self,
+        group: Option<&str>,
```

Comment:
My personal preference is to have two separate APIs for single-valued and multi-valued measurements instead of this `Option` group argument.